Registry per Cluster prototype

nachocano opened the pull request:
## Proposed Changes

- Adding EventType CRD with cluster scope.
- Configuring few ingress policies in Broker.
- RBAC permissions for ingress to retrieve eventtypes (not entirely working at the cluster level).

nachocano committed:
Adding EventType CRD

nachocano committed:
Update to type

nachocano committed:
Updates

nachocano committed:
updating eventtypes to event-types

nachocano committed:
Ingress policy

nachocano committed:
Updating environment variables

nachocano committed:
yamls to give event-type permissions

nachocano committed:
Adding namespace to ingress and reconciling service accounts in ns.

nachocano committed:
More changes

nachocano committed:
Merge remote-tracking branch 'broker/broker-new-model' into registry

# Conflicts:
#	cmd/controller/main.go
#	config/500-controller.yaml
#	pkg/reconciler/v1alpha1/broker/broker.go
#	pkg/reconciler/v1alpha1/namespace/namespace.go

nachocano committed:
Service account in namespace

nachocano committed:
Updating events registered

nachocano committed:
Updating origin to source.

nachocano committed:
to newer version

nachocano commented:
@Harwayne
@grantr

As FYI, an initial version of the registry I've been working on..
Will take a look at how events are being sent by our current sources.

nachocano commented:
FYI @akashrv

nachocano committed:
moving the ingress policy to the broker itself...
should be an object instead of a string.

grantr commented on pkg/broker/ingress.go:
```diff
@@ -0,0 +1,87 @@
+
+func (policy *AllowRegisteredPolicy) AllowMessage(namespace string, message *provisioners.Message) bool {
+	eventType := &eventingv1alpha1.EventType{}
+	name := message.Headers["Ce-Eventtype"]
```
This should probably be a constant.

According to https://github.com/cloudevents/spec/blob/master/http-transport-binding.md, the header name in the HTTP transport is `ce-type`. Is this transformed to `Ce-Eventtype` somewhere else, perhaps in the CloudEvents SDK?

nachocano replied:
Yes, I agree. Should be some constant probably.

I asked Scott, he said that ce-type is spec version 0.2. For 0.1 is called Ce-Eventtype.

nachocano replied:
Actually, Scott suggested we should use the SDK to decode the message. 
I'll do those changes... We need to do it as well for the filtering in triggers...

grantr commented:
This seems fine to me, but I'll let @harwayne do the final review. Thanks @nachocano!

nachocano commented:
@Harwayne
@grantr 

Will add changes to use the SDK to decode the event... so we don't have to care about those constants... Will add that change in the trigger filter prototype we have as well..

nachocano commented:
I take that back. Seems that is a bigger change than I thought. We have the Message struct everywhere. Was wondering whether we should cloudevent.Event. 
Will talk to you guys in person, will leave it as is for now, and try to understand better the sources.

nachocano committed:
making header a constant. Should use the SDK instead.

nachocano committed:
Adding dummy change to trigger event

nachocano commented:
@Harwayne
@grantr 

I was able to send events from four different sources (k8s, cronjob, gcppubsub, github). Attached are the yamls of the sources. At the end of each file you can see received events in the dumper.
For github I was listening to pull_request and push. I created a PR, and then did a push in the same PR branch, but the event triggered was still a pull_request, but with a different action (synchronize). You can see it in the log.
All are Cloud Events v0.1. I guess we are converting them to that spec version somewhere. If we will keep on doing that, then having that constant for Ce-Eventtype and Ce-Source might be good enough. Otherwise I think we should try to include Scott's cloudevent SDK to parse the message.

[eventsFromSources.zip](https://github.com/nachocano/eventing/files/2917512/eventsFromSources.zip)

nachocano committed:
Using cloudevents sdk but still needs to be updated.
We should change our APIs to receive Events instead of Messages.

nachocano committed:
Merge pull request #3 from nachocano/test_pr

Adding cloudevent go sdk

nachocano committed:
Adding cloudevents-sdk library

nachocano committed:
Invalid eventtype

nachocano committed:
Merge remote-tracking branch 'broker/broker-new-model' into registry

nachocano committed:
adding type as authoritative

nachocano committed:
Should actually compare to lowercase

nachocano committed:
listing all event types... we should use some sort of selectors.
label selectors won't work on spec.type as they may not conform to the
label format.
field selectors doesn't seem to work for CRDs beyond metadata.name and
namespace.

nachocano committed:
Might not need origin after all, but adding it here

nachocano committed:
adding origin to yaml as well

nachocano committed:
eventtypes instead of event-types to maintain consistency.

nachocano committed:
updating example

nachocano committed:
Merge branch 'registry' of github.com:nachocano/eventing into registry

n3wscott commented on cmd/broker/ingress/main.go:
```diff
@@ -156,3 +188,23 @@ func (r *runnableServer) Start(<-chan struct{}) error {
 }
+
+// TODO this should be removed once we update the interfaces and start using cloudevents.Event instead of Message.
+func cloudEventFrom(m *provisioners.Message) cloudevents.Event {
```
oh man you should use the cloudevents lib: https://github.com/cloudevents/sdk-go

nachocano replied:
Yes, I put that in the TODO. 
BTW how do I convert the provisioners.Message to a cloudevents.Event if I don't inspect the headers? 
Adam was working on changing the API to not receive a Message as an arg, but a cloudevent... I was waiting for that change

nachocano committed:
Removing source

nachocano committed:
cluster-scoped registry to simplify things.

nachocano committed:
updating example

nachocano committed:
removing source

nachocano committed:
updating broker policy to create event types...
updating cloud event SDK version.

nachocano committed:
adding a TODO to update to a cluster-level role binding for the ingress.
If we are planning on using cluster-scoped EventTypes

nachocano commented:
Won't pursue this one. Already have a working version of Registry per namespace, which seems more suitable for our needs in https://github.com/nachocano/eventing/pull/7.
Closing this PR for now
